c/extras/py_ftrl.cc

```diff
@@ -0,0 +1,338 @@
+  gs.add<&Ftrl::get_hash_type, &Ftrl::set_hash_type>("hash_type",
+    "Hashing method to use for strings.\n"
+    "`0` - std::hash;\n"
+    "`1` - Murmur2;\n"
+    "`2` - Murmur3.\n");
+  gs.add<&Ftrl::get_seed, &Ftrl::set_seed>("seed", "Seed to be used for Murmur hash functions.\n");
+
+  mm.add<&Ftrl::fit, args_fit>();
+  mm.add<&Ftrl::predict, args_predict>();
+}
+
+
+PKArgs Ftrl::Type::args_fit(1, 0, 0, false, false, {"df_train"}, "fit",
+R"(fit(self, frame)
+--
+
+Train an FTRL model on a dataset.
+
+Parameters
+----------
+frame: Frame
+    Frame to be trained on, last column is treated as `target`.
+
+Returns
+----------
+    A new Frame of shape `(d, 2)` containing `z` and `n` model coefficients.
+)");
+
+
+void Ftrl::fit(const PKArgs& args) {
+  if (!args[0].is_frame()) {
+    throw TypeError() << "argument must be a frame, not "
+        << args[0].typeobj();
+  }
+  DataTable* dt_train = args[0].to_frame();
+  fm->fit(dt_train);
+}
+
+
+PKArgs Ftrl::Type::args_predict(1, 0, 0, false, false, {"df_test"}, "predict",
+R"(predict(self, frame)
+--
+
+Make predictions for a dataset.
+
+Parameters
+----------
+frame: Frame
+    Frame of shape `(nrows, ncols)` to make predictions for. It must have one 
+    column less than the training dataset.
+
+Returns
+----------
+    A new `Frame` of shape `(nrows, 1)` with a prediction for each row.
+)");
+
+
+oobj Ftrl::predict(const PKArgs& args) {
+  if (fm->get_model() == nullptr) {
+    throw ValueError() << "There is no trained model in place.\n"
+                       << "To make predictions, the model must be either trained first, or "
+                       << "set.";
+  }
+
+  if (!args[0].is_frame()) {
+    throw TypeError() << "argument must be a frame, not "
+        << args[0].typeobj();
+  }
+
+  DataTable* dt_test = args[0].to_frame();
+  DataTable* dt_target = fm->predict(dt_test).release();
+  py::Frame* df_target = py::Frame::from_datatable(dt_target);
+
+  return df_target;
+}
+
+
+void Ftrl::set_model(robj model) {
+  if (!model.is_frame()) {
+    throw TypeError() << "`model` must be a frame, not "
+        << model.typeobj();
+  }
+
+  DataTable* dt_model_in = model.to_frame();
+  const std::vector<std::string> model_cols_in = dt_model_in->get_names();
+  if (dt_model_in->nrows == fm->get_d() && dt_model_in->ncols == 2 &&
+    dt_model_in->columns[0]->stype() == SType::FLOAT64 &&
+    dt_model_in->columns[1]->stype() == SType::FLOAT64 &&
+    model_cols_in == FtrlModel::model_cols) {
+    fm->set_model(dt_model_in);
+  } else {
+    throw ValueError() << "FTRL model frame must have " << fm->get_d() <<" rows, and" <<
+                          "2 columns named `z` and `n`, " <<
+                          "both columns must be of `FLOAT64` type.";
+  }
+}
+
+
+oobj Ftrl::get_model(void) const {
+  DataTable* dt_model = fm->get_model();
+  if (dt_model == nullptr) {
+    throw ValueError() << "There is no trained model available, train it first or set.";
```

Comment:
Why not return `None` instead?

Reply:
Yes, can also do that. For the moment I'm not sure which way is better.

Reply:
Well, it seems to me that returning `None` is exactly the same as telling the user that "there is no trained model available". But it saves you from writing an extra method `is_model_available()`. However, exceptions are best to use for genuine error conditions. In your case having no trained model is a perfectly valid state, so why should it raise an error?

Reply:
I imagine the following use-case: someone created an `Ftrl` object, forgot to train it, and saved the model with something like `fm.model.to_csv("my_model.csv")`. As there is no error, one may think that everything went well, transfers this file somewhere to make predictions and get an error on the other side.

Sure, one can make a check weather or not the returned `model` is `None`, but this check is not always done. 

Anyways, I'm now thinking that may be the `model` should always be present. If it was not trained, it could be just zero-initialized. Then it will be always a valid state.

Reply:
Well, if `fm.model` returns `None`, then calling `.to_csv()` on it is no good :)
On the other hand, if you return a zero-initialized model, then you'll get exactly the situation that you described: someone forgets to train the model, saves it, then makes garbage predictions on the other side.

I guess the question is: could there be a scenario were returning a zero-initialized model is useful? I can't think of any. And it won't let you answer a simple question: was the model trained already or not?

Reply:
Sure, this is something I've been thinking about too. I meant It will always be a valid state internally, once trained we can set up some `was_trained` flag, and then we control what to return as an `fm.model` to the user. May be `None` is not a bad idea, but I would also like some kind of warning.